torch/csrc/Storage.cpp

```diff
@@ -84,15 +84,124 @@ PyObject* THPStorage_NewWithStorage(
   return obj;
 }
 
+// This object is used as the `ctx` argument for DataPtr to implement a shared
+// DataPtr. Normally, a DataPtr is unique, but we use this custom context and
+// the deleter function below to make the DataPtr act like a non-unique
+// DataPtr. This context object holds onto an inner context and deleter
+// function which handle the actual deletion of the data when the refcount
+// reaches 0.
+//
+// This shared DataPtr feature is only used when storages are shared between
+// multiple Python interpreters in torch::deploy (MultiPy). Before storages had
+// PyObject preservation, interpreters could just share the same StorageImpl
+// instance. But now a StorageImpl can only be associated with one interpreter
+// in order to properly manage a zombie PyObject. So we share storages across
+// Python interpreters by creating a different StorageImpl instance for each
+// one, but they all point to the same data.
+struct RefcountedDeleterContext {
+  RefcountedDeleterContext(void* other_ctx, c10::DeleterFnPtr other_deleter)
+      : other_ctx(other_ctx, other_deleter), refcount(1) {}
+
+  void incref() {
+    refcount++;
+  }
+
+  std::unique_ptr<void, c10::DeleterFnPtr> other_ctx;
+  size_t refcount;
```

Comment:
Does this need to be atomic? 🤔  Maybe it's always GIL protected??

Reply:
No it doesn't seem to be GIL protected, since multiple different interpreters will share the same instance, so I think you're right that it needs to be atomic